### ISSUE TITLE
Performance issue on Linux AppImage v 1.2.0

### ISSUE THREAD

**Issue by flesciamento** [synfig/synfig#314]
**Synfig version & platform**:
Synfig 1.2.0 (last stable release) on Ubuntu 16.04

**Issue description**:
Synfig works very slowly even compared with the previous stable version (installed with deb packages).

**Comment by morevnaproject** [comment]
Please provide a testing file to compare with.

**Comment by boubouh** [comment]
Yes i think synfig  should have .deb copy for newes build ... its more stable

**Comment by flesciamento** [comment]
Sorry for the late! @morevnaproject do you mean a Synfig file? Well, the problem shows up with every files, but if you want to compare, I think I can send you this: 
[Scena 03 v3.sifz.zip](https://github.com/synfig/synfig/files/743694/Scena.03.v3.sifz.zip)

It's a file with a lot of vector stuff, but in the 1.0.2 version I can work (slowly), whereas in the last stable version (1.2.0) every change (even the zoom out!) takes a lot of time.

Even the preview takes the double of time to load every frame.

**Comment by caryoscelus** [comment]
Yeah, i can confirm this with this file. While on small files new rendering engine seems to work faster (1.5-3x), it chokes on such a heavy one (more than 10x slower). Obviously my measuring wasn't entirely correct due to different compilation flags and compilers, but i think that doesn't change much.

**Comment by caryoscelus** [comment]
However, disabling the tree (which has those blurry things that are btw rendered differently) brought a huge speed up and new engine is only about 3 times slower.

**Comment by blackwarthog** [comment]
Wow! It's hard benchmark.

Problem raises when too many regions (~1000) placed at one surface. Renderer wastes a lot of time to determine task dependencies, that required for multithreading.

I've fixed this stage of rendering.  Speed should be increased by ~50%.

**Comment by morevnaproject** [comment]
This fix now available in the latest development version - https://morevnaproject.org/2017/02/04/synfig-studio-1-3-1-development/

**Comment by morevnaproject** [comment]
I have renamed issue, since it is not really related with AppImage (it was the same with DEB and RPM packages).

@flesciamento  I have noticed you produce artwork in Inkscape and then import in Synfig. I can recommend to use boolean operations of Inkscape to merge multiple similar elements into single vector shape. That would speedup your operations.

**Comment by flesciamento** [comment]
@morevnaproject thanks for the tip! :)
I tried this last version and, in my pc, the problem is still here: for rendering every frame it takes about 4~5 secs, while in the 1.0.2 version it takes 2~3 secs! Plus, zoom-out and zoom-in freezes everything for a couple of seconds, while in the 1.0.2 zoom is instantaneous!

**Comment by ricolandia** [comment]
Sadly, I have the same problem here with many other projects. They open and run better in the old 1.0.2 version. I've tested the 1.3.1, and the performance is not so good as 1.0.2 too.

Ubuntu 16.04.1 - 64bit - and - Elemenary OS ( based on ubuntu 16.04 )
Both with Core i5
RAM 4 Gb and 8 Gb

**Comment by Svarov-RZM** [comment]
This issue is probably very hard to debug due to multi threading.
Posted a 30$ bounty to motivate developers.
https://www.bountysource.com/issues/41465340-performance-issue-with-complex-vector-artwork

**Comment by morevnaproject** [comment]
Yes, the optimization  of code takes more than we expected. @blackwarthog makes significant modifications to architecture to eliminate this issue. Thank you very much for support, this is much appreciated!

**Comment by Svarov-RZM** [comment]
Plus 20$ to bounty, 50$ in total.

**Comment by blackwarthog** [comment]
Thank you very much!

For now i'm working on special type of resources which allows to represent image data in different formats (Software or OpenGL for ex.) with auto-conversion between them. With this improvement we can to remove image-conversions from "optimization"-stage of rendering and increase performance of this stage.

Also it's open a way to next improvement. Cobra-engine receives tasks for rendering unspecialized - not software nor opengl. And engine makes decision and selects one of available implementations of these tasks (SW or GL hardware). But each implementation of rendering task can work only with images in special format - software tasks can work with images in RAM, OpenGL task can work oly with images in video-memory. This restriction  Using of auto-converted images (aka textures or surfaces) allows to

**Comment by blackwarthog** [comment]
Thank you very much!

Work still in progress and I'll try to explain here a technical details of this issue.

For now i'm working on special type of resources which allows to represent image data in different formats (Software or OpenGL for example) with auto-conversion between them. With this improvement we can to remove image-conversions from optimization-stage of rendering and increase performance of this stage.

Also it's open a way to next improvement. Cobra-engine receives abstract (unspecialized) tasks for rendering - not software nor opengl. And engine makes decision and selects one of available implementations of these tasks (SW or GL hardware). But each implementation of rendering task can work only with images in special format - software tasks can work with images in RAM, OpenGL task can work only with images in video-memory. So we cannot to create and assign target and source images for unspecialized task. No images - no coordinates of target and source rectangles.

In current code for each implementation of each task we use it own function for specialization which makes three things:
1. convert unspecialized task to specialized;
2. create and assign target and source images in special format;
3. calculate coordinates of target and source rectangles.

With auto-convertable images we don't need to care about format of images and we can just convert unspecialized task into specialized by generic function using preconfigured table of conversion. And we can to calculate coordinates before specialization for abstract task and don't need to duplicate this code for each task implementation. All of this will significantly improve a rendering performance and make future development much easier.

**Comment by morevnaproject** [comment]
The main idea of @blackwarthog is to reduce the number of optimization operations, which are applied before the actual rendering takes place. Currently the number of optimization operations depend on the number of objects and this is the reason why vector artwork with many elements (this issue) is slow. Reducing number of optimizations in the way explained above should reduce this bottleneck.

**Comment by zozorg** [comment]
baby bounty added ($5)

**Comment by Svarov-RZM** [comment]
Thanks for clarification, guys!

I do understand that code refactoring/optimization can often be hard and somewhat boring. So take your time, I'll try to increase the bounty a little more with my next payment.

**Comment by Svarov-RZM** [comment]
Plus another $20, total: $75

**Comment by Svarov-RZM** [comment]
Plus $30 to bounty.
It's $105 in summary and this is where I stop.

**Comment by morevnaproject** [comment]
@Svarov-RZM Thank you very much for support! Right now we are working on releasing version 1.3.5 (with a long due features). After that we will get back to solving issue. Thank you again!

**Comment by blackwarthog** [comment]
plans at paper seems to be ready...

![sam_9413](https://user-images.githubusercontent.com/5143743/33163381-e8f60cae-d060-11e7-99cc-b26f9d418758.JPG)
![sam_9414](https://user-images.githubusercontent.com/5143743/33163382-e92a5928-d060-11e7-8407-1e4aa7f56792.JPG)
![sam_9415](https://user-images.githubusercontent.com/5143743/33163383-e9599896-d060-11e7-8cb4-07010a6a186b.JPG)
![sam_9416](https://user-images.githubusercontent.com/5143743/33163384-e9a6c1d4-d060-11e7-89af-39369015805b.JPG)
![sam_9417](https://user-images.githubusercontent.com/5143743/33163385-e9f841e4-d060-11e7-8f32-1a7fe8b01230.JPG)
![sam_9418](https://user-images.githubusercontent.com/5143743/33163386-ea66dabe-d060-11e7-8eb7-e51e21f1d257.JPG)
![page1](https://user-images.githubusercontent.com/5143743/33163387-ec01d9d2-d060-11e7-8bfd-f7b84d2f2bee.JPG)
![page2](https://user-images.githubusercontent.com/5143743/33163388-ec505fe4-d060-11e7-905b-6fd8ad788396.JPG)
![page3](https://user-images.githubusercontent.com/5143743/33163389-ec7dd898-d060-11e7-8f9c-1dfab13bc601.JPG)
![page4](https://user-images.githubusercontent.com/5143743/33163391-ed316aca-d060-11e7-8a96-2c80c1d4cf7f.JPG)
![page5](https://user-images.githubusercontent.com/5143743/33163393-ed7dcd5c-d060-11e7-99b8-ed3baddbf7ff.JPG)
![page6](https://user-images.githubusercontent.com/5143743/33163394-eddab31e-d060-11e7-94b4-8fc5dc14b4d4.JPG)
![page7](https://user-images.githubusercontent.com/5143743/33163396-ee26b534-d060-11e7-853f-ad07c5f3e0a7.JPG)
![page8](https://user-images.githubusercontent.com/5143743/33163397-ee70531a-d060-11e7-8ab8-e0c34b119f31.JPG)

**Comment by Svarov-RZM** [comment]
Posted another $25, $130 in total.

**Comment by morevnaproject** [comment]
@Svarov-RZM Thank you for support! And apologies for delaying with this feature.  Unfortunately, @blackwarthog is our only developer at the moment and he got many tasks to do at one. I am still struggle to get second developer (to handle bugfixes and other stuff besides main development), but that still take some time.

**Comment by blackwarthog** [comment]
work in progress

https://github.com/blackwarthog/synfig/tree/optimize
https://github.com/blackwarthog/synfig/commit/75dea815e3fe979c43da98c9cf1ca3d827d8f9bd

**Comment by Svarov-RZM** [comment]
No worries, guys. Take your time.
I am going to increase bounty a little more to at least $150 in the end of this month.

**Comment by Svarov-RZM** [comment]
+$25 as promised, total $155.

**Comment by Svarov-RZM** [comment]
+$50, total $205.
God, render needs work. I recently had some problems with gradients and distortions on a FullHD scene that weren't present in 1.0.2 version. I hope when this issue gets closed it'll be easier to fix remaining bugs.

**Comment by morevnaproject** [comment]
Sample file from https://github.com/synfig/synfig/issues/314#issuecomment-276556145, but without blurs - 
[Scena 03 v3-noblur.sifz.zip](https://github.com/synfig/synfig/files/2119358/Scena.03.v3-noblur.sifz.zip)

**Comment by morevnaproject** [comment]
We have new re-optimized rendering engine merged into master branch now and ready for testing. Much thanks to @blackwarthog for his epic work and dedication!

**Comment by Svarov-RZM** [comment]
Hey, I can see there's a lot of progress.
I can't test the Master branch myself, because I am a Windows guy and not sure if I'll be able to compile Synfig on it.

Anyway, I added another small bounty: $15, total: $220.

**Comment by morevnaproject** [comment]
@Svarov-RZM Thank you very much for your support! Please check this testing build with latest changes - https://dev.icystar.com/downloads/SynfigStudio-1.3.10-testing-18.09.04-win64-eba13.zip

**Comment by Svarov-RZM** [comment]
@morevnaproject 
Thanks for the build!
I have some ambiguous feelings about it. On some projects it works faster, but on others it can still be slower than 1.0.2 version. Oh, and there are also some strange decisions or maybe bugs. OK, here's what I've got:

1. Timebar preview in the bottom (green line) is not very useful. It takes A LOT of time to render on big projects and because of that eats 100% of one CPU core most of the time. Also, it resets when you zoom in/out or show/hide layers. It looks cool, yeah, but I personally want to disable it. Maybe provide an option to disable this live preview?
2. Also, rendering of the timebar preview does not stop when you start actual preview and now you have two rendering processes that slow each other down.
3. Flickering on editing. Try to edit some shape in animation mode; the whole canvas will be cleared first and after rendering you get your updated image. It's very hard to focus on editing now because canvas is constantly flickering/flashing. 
4. There's some bug with standalone waypoints (those not connected to any keyframe) I think. See example animation below.

I've attached one of my animation where I experimented with explosion effects.
You can see rendered result here: https://www.youtube.com/watch?v=GCLoMMtByOo
Project is pretty complex and involve a lot of effects (blurs and distortions) so it makes a good test subject. Open this project and try to preview first 5 seconds. By some reason, preview renders it with some lags and 'jumps' like if it renders the project randomly.

[Bomb4.zip](https://github.com/synfig/synfig/files/2408970/Bomb4.zip)

**Comment by morevnaproject** [comment]
Hello @Svarov-RZM! Thank you for the feedback. Here are some clarifications:

> On some projects it works faster, but on others it can still be slower than 1.0.2 version.

This is because we haven't ported all layers to new engine yet. With this build we have addressed performance issues with complex vector artwork - i.e., artwork which is purely constructed from vectors (outlines, advanced outlines and regions). The original example of such artwork provided in https://github.com/synfig/synfig/issues/314#issuecomment-276556145 - a file with thousands of region/outline layers.

So, if you will try to test with pure-vector artwork, then this build should be faster than 1.0.2.
Hint: as of today, advanced outline layer  can be optimized even more.

For other types of layers this is not guaranteed, because they are not all ported yet. Synfig have many different layers with specific functionality, and each of them should be treated specifically.
For example, we haven't ported distortion layers yet (which are heavily used in example you provided), so they are still very slow.

> Timebar preview in the bottom (green line) ... It looks cool, yeah, but I personally want to disable it. Maybe provide an option to disable this live preview?

Yes, we have this in our plans ^__^ - https://github.com/synfig/synfig/issues/611

> it resets when you zoom in/out or show/hide layers

One of our next steps is to implement a resolution-based rendering mode. I.e. the workarea rendering will be linked to canvas dimensions. This is how it works with vector graphics in Krita - you  can work with vector elements there. But when you zoom-in  -  you see pixels. That's because rendering of vector elements  is linked to document resolution (dimensions), not to current view.
This is highly improving performance and removing the need to re-render cache on every zoom/pan.

> rendering of the timebar preview does not stop when you start actual preview and now you have two rendering processes that slow each other down.

We plan to remove preview window, since its functionality is replaced by timeline cache (it does exactly the same thing).

> Flickering on editing.

I haven't been able to reproduce that.  Can you please record a video about this issue?

> There's some bug with standalone waypoints

Can you please attach a video for that as well? I wasn't able to find out which exact rpoblem you are referring to.

Thank you!

**Comment by morevnaproject** [comment]
> Ah, I see. So "Performance issue with distortion effects" is inevitable?

No, it is not inevitable. ^_^ What I want to say is that it is a different task comparing to "vector artwork optimization" and needs a special treatment. We will surely pay attention to optimization of other layers as soon as get done with this task. ^__^

> Wait a second. Timeline cache is that little preview thing in the bottom, right? 

Yes.

> But the current preview window allows you to set quality/start-end and even zoom-to-fit/200% zoom which I use all the time. Are you going to drop this functionality?

**- start/end**

You can set start/end using playback range - https://www.youtube.com/watch?v=xwbB0Q20Myo

In the latest code all frames from the playback range are rendered with top priority (https://github.com/synfig/synfig/commit/8377b514088565a9c0bfbc5b0a157ed0e2aaf695). (This change is not available in binary builds yet).

**- quality**

You can use "Low res" button at the top of the workarea - it works the same way as Quality parameter of preview (reducing the resolution).

![screenshot_001](https://user-images.githubusercontent.com/332868/46615563-ae754080-cb42-11e8-83c1-3c3ba37e74f0.png)

 **- zoom-to-fit/200% zoom**

There are zoom buttons at the bottom left corner of the workarea to control zoom -

![screenshot_003](https://user-images.githubusercontent.com/332868/46615679-0ad86000-cb43-11e8-8d58-d923d68b802c.png)

Please let me know if you have any other missing functionality of Preview Window, so we can discuss that. ^__^

> Sure thing: https://1drv.ms/v/s!Aqn9-DWzZKfcgUditYZixzuC8qwt
Pay attention how the canvas gets cleared on EVERY change and I was editing just a simple region. On a more complex scenes this flickering is even more annoying because rendering needs more time to finish.

Ah, I see now! Yes, this is very annoying issue and @blackwarthog already fixed it.
The next build will include this fix.

**Comment by morevnaproject** [comment]
@Svarov-RZM Please check this fresh build (fixes flickering bug) - https://dev.icystar.com/downloads/SynfigStudio-1.3.10-testing-18.10.09-win64-12dbc.zip

**Comment by Svarov-RZM** [comment]
I think I get it. You want to integrate functionality of Preview window into main workspace, correct? Seems reasonable. There are buttons to play animation in main window but it lags when playing. I just hope that it gets done right, because now it's a bit of a mess.

Regarding new build:
There's an issue with rendering outline's ending cusp. See the picture below and also attached sifz file.
Left — 1.3.10 version, right — 1.0.2.
![cusps](https://user-images.githubusercontent.com/28569334/46933797-30db9400-d05e-11e8-924f-2589a6a57358.png)
[Cusps.zip](https://github.com/synfig/synfig/files/2477624/Cusps.zip)

**Comment by morevnaproject** [comment]
> I think I get it. You want to integrate functionality of Preview window into main workspace, correct?

Exactly! ^__^

> There are buttons to play animation in main window but it lags when playing. I just hope that it gets done right, because now it's a bit of a mess.

Yes, we are working on that now.

BTW, here is a fresh build, which allows to [disable background rendering](https://www.patreon.com/posts/weekly-report-8-22048380) - https://dev.icystar.com/downloads/SynfigStudio-1.3.10-testing-18.10.18-win64-defe1.zip

> There's an issue with rendering outline's ending cusp.

Can you please submit this test file to our regression test suite? - https://github.com/synfig/synfig-tests-regressions/issues/3
Having this example there will help us to track the regressions and avoid them in the future!
Please check this discussion for details about  regression test suite - https://github.com/synfig/synfig/issues/634

Thank you! ^__^

**Comment by Svarov-RZM** [comment]
> BTW, here is a fresh build, which allows to [disable background rendering](https://www.patreon.com/posts/weekly-report-8-22048380)

Looks good. With background rendering disabled the program is more responsive and lags less when you work with current frame.
Tested the program a bit harder and I noticed a difference in how 'Bevel' level is rendered. Seems like 'Angle' parameter is calculating differently from 1.0.2 version. I'll make a comparison later.



> Can you please submit this test file to our regression test suite?

How can I do that? Should I use git for that? Sorry, I only used this tool to get sources and 'make install' them, I never contributed to anything using it. Can you please give me some advice on where to start?

**Comment by morevnaproject** [comment]
@Svarov-RZM My apologies for not responding in time - the message got lost in the stack of issues. ^__^" Thank you for submitting pull request, I have included it into our test suite now. Will be working on the fix.

**Comment by Svarov-RZM** [comment]
@morevnaproject Can I ask what is the current state of this issue? It seems to be solved... or am I wrong?
I mean there's an outline issue but it's a general bug, not a performance issue.

So, what's holding this one from closing?

**Comment by morevnaproject** [comment]
@Svarov-RZM The optimizations are applied for Region and Outline layers. But Advanced Outline layer is not optimized yet. @blackwarthog  wants to do some optimizations related with segments of advanced outline.

**Comment by Svarov-RZM** [comment]
Got it. Posted another $15 so it's $250 total.
God, that bountysource site is really buggy...

**Comment by morevnaproject** [comment]
Thank you!

**Comment by morevnaproject** [comment]
Advanced Outline is rewritten as per commit 573f832f5c0c2c9eecc74b811ae9e2943dbae3a2,

Warp Layer now also  optimized (commits cf1df92b252699989e3c36ce4406259a1e40ddd3, fc7a046ad041978bf4071d1e15e6dc2e64176d4c).

Much thanks to @blackwarthog!

Can anyone check performance with Warp Layer?
It would be nice to do some tests comparing 1.2.0, 1.3.12 and latest master.

**Comment by morevnaproject** [comment]
I have tested new Warp Layer - https://forums.synfig.org/t/please-help-with-testing-of-new-warp-layer/10687

**Comment by morevnaproject** [comment]
@flesciamento @boubouh @ricolandia @Svarov-RZM 
The updated version is available for released now - https://www.synfig.org/2020/03/17/synfig-studio-1-3-13-released/

It doesn't includes Warp Layer optimizations, but it has all other optimzations for vector artwork.

Can you please test it and let me know if issue is resolved? ^__^

**Comment by flesciamento** [comment]
Definitely better! I tested it with the same file and other files. Zoom in and zoom out are instantaneous like the 1.0.2 version and everything go still slow, but much smoother and faster than before. Good work!

**Comment by Svarov-RZM** [comment]
@morevnaproject 
OK, it took me a while to understand why half of my projects appeared messed up in some weird way and I finally got it - there's a pretty big bug in the latest builds:

* Spline vertex converter calculated inverted. Value 0.25 would appear like 0.75 in Synfig 1.0.2
* Same stuff with width points on Advanced outlines
* Tangents on Advanced outline are also inverted

This causes compatibility issues with all projects that were done in previous versions. Here's a combined screenshot that shows the problem (pay attention to red dot and direction of tangents):
![SplineVertexBug](https://user-images.githubusercontent.com/28569334/77100141-33e26d00-6a26-11ea-8de2-de4e6bbf7485.png)

The source file is also attached.
[SplineVertexBug.zip](https://github.com/synfig/synfig/files/4356256/SplineVertexBug.zip)

P.S. Considering the speed of outlines/regions it's fine. I don't really have a lot of complex vector work to test though, it mostly simple shapes plus hella lot distortions/blurs/warps/other effects and that is where I am suffering.

**Comment by morevnaproject** [comment]
@Svarov-RZM Owww, that's serious! I can reproduce the issue. Thanks for taking your time on digging into issue and providing an example!

I also tested versions 1.2.2 and 1.3.11 - they both give same result as 1.0.2.

@ice0 I consider this as blocker to 1.4.0 stable release.

@blackwarthog any ideas how to fix this?

**Comment by morevnaproject** [comment]
Interesting: same light-blue curve (advanced outline) is rendered correctly if it is not connected to shape. I have created an example by copy-pasting it to separate file (in Synfig 1.0.2) - https://gitlab.com/synfig/synfig-tests/-/raw/master/rendering/sources/layers/advanced-outline/advanced-outline-002.sif?inline=false

**Comment by rodolforg** [comment]
I did a git bisect and found out the problem is this commit:
a90d014c56f0cbee21c2ab0ef68936b9e84c49b1

**Comment by morevnaproject** [comment]
Thank you @rodolforg, this helps a lot!

In relation to that... can you please find which commit is triggering #1014? (I suspect this could be 573f832f5c0c2c9 or later)

**Comment by rodolforg** [comment]
Do you know the last working version and a synfig testing file? It helps a lot. :)

**Comment by morevnaproject** [comment]
@rodolforg I will try to make some! ^__^

I have talked with @blackwarthog and he pointed out that the problem comes because of differences how point positions calculated in old and new implementation of outline.

The difference takes place in looped case only.

First, let's look how it works for unlooped outline.
Here everything is simple - 1st vertex is corresponding to position "0". And last vertex is corresponding to position "1".

![Изображение-1](https://user-images.githubusercontent.com/332868/77514914-3d4a5a00-6eaa-11ea-896e-19e5ad840310.jpeg)

Here is how it works for looped outline in old implementation:
The position "0" (start of outline) is aligned to LAST vertex of outline.

![Изображение-2](https://user-images.githubusercontent.com/332868/77514921-41767780-6eaa-11ea-8d84-ea579949f6bb.jpeg)

@blackwarthog found this counter-intuitive, so in new implementation he made position "0" aligned to FIRST vertex of outline:

![Изображение-3](https://user-images.githubusercontent.com/332868/77514931-463b2b80-6eaa-11ea-9013-2e26c5f75eca.jpeg)



Below you can see a modification of example given by @Svarov-RZM to illustrate how it works for looped outline with 3 vertices (I put numbers near vertices for better understanding):

Old version:
![screenshot_004](https://user-images.githubusercontent.com/332868/77515609-661f1f00-6eab-11ea-9365-6d2b4f34070b.png)

New version:
![screenshot_005](https://user-images.githubusercontent.com/332868/77515616-6a4b3c80-6eab-11ea-802e-7bdd23d07bb8.png)

So, in new version the position on outline is shifted by **length of segment between first and last vertex**. 

The value of this length is depends on  the value of "Homogenous" parameter:

![screenshot_006](https://user-images.githubusercontent.com/332868/77515765-b8604000-6eab-11ea-960e-a51d8a896893.png)

[a] If Homogenous parameter is false, then length of offset is equal to 1/vertex_count.
[b] If Homogenous parameter is true, then length of offset is equal to real length of segment between first and last vertex. And if outline is animated that could be a dynamic value!

So, I think we should revert to old behavior: **in looped outlines the position "0" should be corresponding to last vertex.**

**Comment by morevnaproject** [comment]
@blackwarthog said that he will not be able to work on this issue in nearest time.

And since this issue is blocking next stable release, I have two options:

1. Wait for someone to solve it (and that means delay stable release for undefined time).
2. Rollback ALL modifications related to refactoring of outline in 1.3.x branch and release without new code for outline. The code will remain in maser branch and will be available in 1.5.0 (next development release). At the same time, we will have stable release shipped with all new functions that people made for more than one year.

**Comment by morevnaproject** [comment]
> Do you know the last working version and have a synfig testing file to share with me? It helps a lot. :)

@rodolforg I accidentally found guilty commit. ^__^ It is confirmed to be 573f832 

Test file - 
[outline-test.sif.zip](https://github.com/synfig/synfig/files/4380017/outline-test.sif.zip)

In previous commit it looks like this -
![screenshot_007](https://user-images.githubusercontent.com/332868/77521950-0c702200-6eb6-11ea-95be-93871759485c.png)

If you compile with 573f832, then (besides other artifacts) you can see that outline width is not corresponding to width points.

**Comment by morevnaproject** [comment]
I found even simpler test to demonstrate the problem reported by @Svarov-RZM:

[outline-test-2.sif.zip](https://github.com/synfig/synfig/files/4380047/outline-test-2.sif.zip)

In Synfig 1.0.2:
![screenshot_008](https://user-images.githubusercontent.com/332868/77522396-c798bb00-6eb6-11ea-98b2-af15d601015a.png)

In Synfig master:
![screenshot_009](https://user-images.githubusercontent.com/332868/77522442-d7b09a80-6eb6-11ea-85ec-a12af250afbc.png)

(note changes in advanced outline on the left)

**Comment by morevnaproject** [comment]
Wait... it seems I found solution... What if we change order of vertices
for all looped outlines when file is loaded?

пн, 23 марта 2020 г. в 21:40, rodolforg <notifications@github.com>:

>
> --
http://morevnaproject.org/

**Comment by morevnaproject** [comment]
@Svarov-RZM The issue you described in https://github.com/synfig/synfig/issues/314#issuecomment-601336943 is fixed now. The fix will be shipped in version 1.3.14. Thank you for reporting the issue!

**Comment by Svarov-RZM** [comment]
@morevnaproject @rodolforg  Thanks guys! I'll be sure to check it when Windows build is available and report back.

**Comment by morevnaproject** [comment]
The fix is now available in version 1.3.14 ^__^ - https://www.synfig.org/2020/05/07/synfig-studio-1-3-14-released/

**Comment by Svarov-RZM** [comment]
@morevnaproject 
I tested the new development version and all seems to work as expected. At least I didn't notice anything wrong with my own vector work.

Side question: is there a list of layers that are not yet ported to a new engine? I would like to keep a close look on the progress, especially distortion layers.

**Comment by IcedQuinn** [comment]
There is still an active bounty open for this. It looks like the issue has been solved or at least partly so. Should this ticket (and bounty) remain open?

**Comment by Svarov-RZM** [comment]
> Should this ticket (and bounty) remain open?

Well, Synfig handles vector graphic much better after all the work that was done in this issue, so I guess it can be closed but what about the bounty? In order to give it to someone, that someone must submit a claim on the Bountysource site and no one did.

To be honest, I am not very kin on giving the bounty to the original developer (blackwarthog). He didn't send request, didn't write any kind of report/summary, just moved to other projects without giving any information. He clearly doesn't care, so why should I? I am OK on giving bounty money to Konstantin though, he's a long-time manager of Synfig, I am sure he will handle it appropriately.

**Comment by zozorg** [comment]
Speaking of bounties - bountysource doesn't update crypto prices, note that bounties for

https://app.bountysource.com/issues/61683574-make-onionskin-work-in-the-same-way-as-in-opentoonz
https://app.bountysource.com/issues/54370243-keyboard-shortcuts-collapse-tree-and-create-spline

... aren't ~$100 as displayed, but rather $450 and $570. In case someone is interested.

**Comment by Svarov-RZM** [comment]
All right, I decided to call off the bounty.
The original developer is gone, goals aren't set properly and basically no one cares.
So I will try to contact bountysource for refund and if successful, relocate the money to other Synfig issues.

P.S. I feel that this is a bit of a dirty move, but again, there's no progress.
If you think there's a better way to handle this, do tell.

**Comment by tcuong53-cell** [comment]
I'm working on a fix for this issue. Will submit a clean PR shortly!

**Comment by tcuong53-cell** [comment]
I'm working on a fix for this issue. Will submit a clean PR shortly!

**Comment by Svarov-RZM** [comment]
I am closing this. The root case was solved long time ago.
Here are some benchmarks on my machine (cpu Intel Core i5 650) using the test file attached in the beginning of this thread:
* Synfig 1.0.2 version: renders 1 frame in ~8.2 seconds
* Synfig 1.2.2 version: renders 1 frame in ~9.6 seconds
* Synfig 1.5.3 version: renders 1 frame in ~4.0 seconds

So the performance speed is 2 times faster than in 1.0.2 version on this specific file.

What about bounty? The Bountysource site is dead. I wrote them several mails trying to get a refund but no luck, they just don't respond. So there's no reason to keep this issue open (which also attracts a lot of bots).
